### DISCUSSION TITLE
improve HTML-forms

### DISCUSSION THREAD

**Discussion by Alken0** [rwf2/Rocket#1938] [Ideas]
Hey, 
I am currently working with HTML-forms but it's not a very pleasant experience...

The only other Web-Framework comparable to Rocket I used so far is Django, that's why I will refer a bit to Djangos approach to this problems. This does not mean Rocket has/should to do it the same way! However it would be more convenient to work with Rocket this way.

Problems:
1. Rust form is not auto-translated into HTML
    - mismatched HTML and Rust code is not detected at compile time/... only through unit tests
2. no auto HTML validation (macro validate does not translate to HTML)
3. for #[post...] handlers: it's hard to auto-return a form
    - even example is kind of boilerplate: https://github.com/SergioBenitez/Rocket/blob/master/examples/forms/src/main.rs
    - what would be nice: let content = form.validate("<template?>")? (returning Template response with errors automatically)
4. custom FormFields-validation is not async
    - pub fn valid(... ) -> bool => I am using this e.g. for data-base request or validate a Path which is async in nature but blocks everything
    - however: this may be not intended in valid? / performance issues because of small async 'threads'? - I am not very experienced in this

Suggestion: autotranslate Form to html
- extend FromForm / FromFormField with HTML translation function?
- extend macro "form" that translates Form to HTML in tera/handlebar?
- extend macro "field(validate" with translation to HTML (regex => <input pattern="....)

Suggestion: easier return of template
I believe this is the normal use-case to return without extra context so it should be supported in an easier way so this should be simpler...
- in handler #[post...] handler { let data = form.validate()?; return form.html() }
    - form.validate() returns on error an HTML response with form containing errors
    - form.html() returns HTML response with form
- in form: define template location #[Template("<tempalte-location>")] struct MyForm...
  

I would also help to extend the framework but would like to at least get some feedback and discussion before implementing something in vain.
Also I have no experience in macro-programming so I don't know if this is even possible to implement...

**Comment by edgarogh** [comment]
This looks a bit out-of-scope for Rocket, from my understanding of the project's philosophy. Rocket is an opinionated web framework, but probably not to the point of deciding how HTML should be emitted based solely on the type of a structure's field. Should HTML5 be used? Should inputs have bootstrap classes attached to them? This is very project-dependant.

However, having experience in Symfony, I agree that very high-level form builders are really great. But this should be its own library, and Rocket is "open" enough for an external crate to be able to implement this kind of things.
